### DISCUSSION TITLE
rego.Function runtime extension json ints/floats are strings

### DISCUSSION THREAD

**Discussion by shortwavedave** [open-policy-agent/community#494] [OPA and Rego]
Hello, I have a custom rego function that validates json input based on some custom logic. Howeverm the `a` term which is a json object always converts numbers to strings. I tried changing the Args type to `types.A` and `types.NewObject` but still the same result. Does anyone have any advice?

```golang
		&rego.Function{
			Name:             "validate_custom",
			Decl:             types.NewFunction(types.Args(types.S, types.S), types.A),
			Memoize:          true,
			Nondeterministic: true,
		},
		func(bctx rego.BuiltinContext, a, b *ast.Term) (*ast.Term, error) {
			var inputData map[string]interface{}
			var schemaData string
			if err := ast.As(a.Value, &inputData); err != nil {
				fmt.Printf("Failed to parse inputData: %v\n", err)
				return nil, err
			} else if err := ast.As(b.Value, &schemaData); err != nil {
				fmt.Printf("Failed to parse schemaData: %v\n", err)
				return nil, err
			}
```

for example, an input like this would erroneously convert the number to a string:
```json
{
"an_integer": 0
}
```
looks like this when I step through and debug:
```json
{
"an_integer": "0"
}

**Comment by johanfylling** [comment]
`ast.As()` and its various sibling functions (e.g. `ast.JSON()`, `ast.ValueToInterface()`) will return numbers typed as `json.Number`, which has `string` as its base type. Either you can convert these to an appropriate number via an appropriate conversion function provided by `json.Number` (e.g. `json.Number("1").Int64()`) as you come across them; or, if this doesn't fit with how you're processing your input and you need to do it all upfront, you could do a json round-trip: 

```
j, err := ast.JSON(operands[0].Value)
if err != nil {
	...
}

bytes, err := json.Marshal(j)
if err != nil {
	...
}

var inputData map[string]interface{}
if err := json.Unmarshal(bytes, &inputData); err != nil {
	..
}
```

Can't claim that's very pretty, though 😅.
